File: lib/extractor/product-extractor.js
/**
 * Product Extractor
 * - 쿠팡 검색 결과 페이지에서 상품 정보 추출
 * - 랭킹 상품 vs 광고 상품 구분
 * - 3-part unique key 사용 (product_id + item_id + vendor_item_id)
 */

class ProductExtractor {
  /**
   * 페이지에서 상품 정보 추출 (브라우저 컨텍스트에서 실행)
   */
  static extractProductsFromPage() {
    const productList = document.querySelector('#product-list');
    if (!productList) return { ranking: [], ads: [], total: 0 };

    const links = productList.querySelectorAll('a[href*="/products/"]');
    const rankingProducts = new Map();
    const adProducts = new Map();

    links.forEach(link => {
      const href = link.href;

      // /vp/products/ 형태만 허용
      if (!href.includes('/vp/products/')) return;

      // product_id, item_id, vendor_item_id 추출
      const productIdMatch = href.match(/\/vp\/products\/(\d+)/);
      if (!productIdMatch) return;
      const productId = productIdMatch[1];

      const itemIdMatch = href.match(/[?&]itemId=(\d+)/);
      const vendorItemIdMatch = href.match(/[?&]vendorItemId=(\d+)/);

      const itemId = itemIdMatch ? itemIdMatch[1] : '';
      const vendorItemId = vendorItemIdMatch ? vendorItemIdMatch[1] : '';

      // 고유 키: product_id + item_id + vendor_item_id
      const uniqueKey = `${productId}_${itemId}_${vendorItemId}`;

      // rank 파라미터 확인
      const hasRank = href.includes('rank=');
      let rank = null;

      if (hasRank) {
        try {
          const rankMatch = href.match(/[?&]rank=(\d+)/);
          if (rankMatch) rank = parseInt(rankMatch[1]);
        } catch(e) {}
      }

      // AdMark 클래스 확인
      const productCard = link.closest('li') || link.closest('div');
      const hasAdMark = productCard?.querySelector('[class*="AdMark"]') !== null;

      const nameEl = productCard?.querySelector('.name') ||
                     productCard?.querySelector('[class*="name"]') ||
                     link;
      const priceEl = productCard?.querySelector('.price-value') ||
                      productCard?.querySelector('[class*="price"]');

      const productData = {
        productId: productId,
        itemId: itemId,
        vendorItemId: vendorItemId,
        uniqueKey: uniqueKey,
        name: nameEl?.textContent?.trim() || '',
        price: priceEl?.textContent?.trim() || '',
        url: href,
        rank: rank,
        hasAdMark: hasAdMark
      };

      // rank 파라미터가 있고 AdMark가 없으면 순수 랭킹 상품
      if (hasRank && rank !== null && !hasAdMark) {
        if (!rankingProducts.has(uniqueKey)) {
          rankingProducts.set(uniqueKey, productData);
        }
      } else {
        if (!adProducts.has(uniqueKey)) {
          adProducts.set(uniqueKey, productData);
        }
      }
    });

    return {
      ranking: Array.from(rankingProducts.values()),
      ads: Array.from(adProducts.values()),
      total: rankingProducts.size + adProducts.size
    };
  }

  /**
   * 중복 체크 (여러 페이지의 결과를 분석)
   */
  static checkDuplicates(results) {
    const allProductKeys = [];
    const pageProducts = {};
    const productDetails = {};

    for (const result of results) {
      if (result.success && result.rankingProducts) {
        const pageNum = result.page;
        const productKeys = result.rankingProducts.map(p => p.uniqueKey);
        allProductKeys.push(...productKeys);
        pageProducts[pageNum] = new Set(productKeys);

        result.rankingProducts.forEach(p => {
          if (!productDetails[p.uniqueKey]) {
            productDetails[p.uniqueKey] = [];
          }
          productDetails[p.uniqueKey].push({
            page: pageNum,
            url: p.url,
            name: p.name,
            rank: p.rank
          });
        });
      }
    }

    const totalProducts = allProductKeys.length;
    const uniqueProducts = new Set(allProductKeys).size;
    const duplicateCount = totalProducts - uniqueProducts;

    return {
      total: totalProducts,
      unique: uniqueProducts,
      duplicates: duplicateCount,
      duplicate_rate: totalProducts > 0 ? (duplicateCount / totalProducts * 100) : 0,
      pageProducts: pageProducts,
      productDetails: productDetails
    };
  }

  /**
   * 중복 정보 출력
   */
  static printDuplicates(duplicateStats) {
    console.log(`\n${'='.repeat(80)}`);
    console.log(`[DUPLICATE CHECK - RANKING PRODUCTS (product_id + item_id + vendor_item_id)]`);
    console.log(`${'='.repeat(80)}`);

    console.log(`Total Ranking Products: ${duplicateStats.total}`);
    console.log(`Unique Products: ${duplicateStats.unique}`);
    console.log(`Duplicates: ${duplicateStats.duplicates}`);

    if (duplicateStats.duplicates > 0) {
      console.log(`\n[WARNING] ${duplicateStats.duplicates}개 중복 상품 발견!`);

      const productCounter = {};
      const allKeys = [];

      for (const pageProducts of Object.values(duplicateStats.pageProducts)) {
        for (const key of pageProducts) {
          allKeys.push(key);
        }
      }

      allKeys.forEach(key => {
        productCounter[key] = (productCounter[key] || 0) + 1;
      });

      const duplicates = Object.entries(productCounter).filter(([_, count]) => count > 1);

      console.log(`\n[DUPLICATE PRODUCTS WITH FULL URLS]:`);
      duplicates.forEach(([key, count]) => {
        console.log(`\n  Key ${key}: ${count}번 등장`);

        const pagesWithDuplicate = [];
        for (const [pageNum, productSet] of Object.entries(duplicateStats.pageProducts)) {
          if (productSet.has(key)) {
            pagesWithDuplicate.push(pageNum);
          }
        }
        console.log(`    Pages: ${pagesWithDuplicate.join(', ')}`);

        const details = duplicateStats.productDetails[key];
        if (details) {
          details.forEach(detail => {
            console.log(`    Page ${detail.page}:`);
            console.log(`      URL: ${detail.url}`);
            console.log(`      Name: ${detail.name.substring(0, 50)}...`);
            console.log(`      Rank: ${detail.rank}`);
          });
        }
      });
    } else {
      console.log(`\n[OK] 중복 없음! 모든 상품이 고유합니다.`);
    }

    return duplicateStats;
  }
}

module.exports = ProductExtractor;

File: lib/extractor/product_extractor.py
"""
Product Extractor (Python)
- Coupang search result page product extraction
- Ranking products vs ads separation
- 3-part unique key (product_id + item_id + vendor_item_id)
"""

import re
from bs4 import BeautifulSoup
from urllib.parse import urlparse, parse_qs


class ProductExtractor:
    """Extract product information from Coupang HTML"""

    @staticmethod
    def extract_products_from_html(html):
        """
        Extract products from HTML content

        Args:
            html: HTML string from Coupang search page

        Returns:
            dict with ranking, ads, total counts
        """
        soup = BeautifulSoup(html, 'html.parser')

        # Find product list container
        product_list = soup.select_one('#productList, #product-list')
        if not product_list:
            return {'ranking': [], 'ads': [], 'total': 0}

        # Find all product links
        links = product_list.select('a[href*="/products/"]')

        ranking_products = {}
        ad_products = {}

        for link in links:
            href = link.get('href', '')

            # Only /vp/products/ format
            if '/vp/products/' not in href:
                continue

            # Extract product_id
            product_id_match = re.search(r'/vp/products/(\d+)', href)
            if not product_id_match:
                continue

            product_id = product_id_match.group(1)

            # Parse URL parameters
            parsed = urlparse(href)
            params = parse_qs(parsed.query)

            item_id = params.get('itemId', [''])[0]
            vendor_item_id = params.get('vendorItemId', [''])[0]

            # Unique key: product_id + item_id + vendor_item_id
            unique_key = f"{product_id}_{item_id}_{vendor_item_id}"

            # Check rank parameter
            has_rank = 'rank' in params
            rank = None

            if has_rank:
                try:
                    rank = int(params['rank'][0])
                except (ValueError, IndexError):
                    pass

            # Check AdMark class
            product_card = link.find_parent('li') or link.find_parent('div')
            has_ad_mark = False

            if product_card:
                ad_mark = product_card.select('[class*="AdMark"]')
                has_ad_mark = len(ad_mark) > 0

            # Extract name and price
            name_el = None
            price_el = None

            if product_card:
                name_el = (product_card.select_one('.name') or
                          product_card.select_one('[class*="name"]') or
                          link)
                price_el = (product_card.select_one('.price-value') or
                           product_card.select_one('[class*="price"]'))

            product_data = {
                'productId': product_id,
                'itemId': item_id,
                'vendorItemId': vendor_item_id,
                'uniqueKey': unique_key,
                'name': name_el.get_text(strip=True) if name_el else '',
                'price': price_el.get_text(strip=True) if price_el else '',
                'url': href,
                'rank': rank
            }

            # Pure ranking products (has rank, no ad mark)
            if has_rank and rank is not None and not has_ad_mark:
                if unique_key not in ranking_products:
                    ranking_products[unique_key] = product_data
            else:
                if unique_key not in ad_products:
                    ad_products[unique_key] = product_data

        return {
            'ranking': list(ranking_products.values()),
            'ads': list(ad_products.values()),
            'total': len(ranking_products) + len(ad_products)
        }

    @staticmethod
    def check_duplicates(results):
        """
        Check for duplicate products across multiple pages

        Args:
            results: List of result dicts with rankingProducts

        Returns:
            dict with duplicate statistics
        """
        all_product_keys = []
        page_products = {}
        product_details = {}

        for result in results:
            if result.get('success') and result.get('rankingProducts'):
                page_num = result['page']
                product_keys = [p['uniqueKey'] for p in result['rankingProducts']]
                all_product_keys.extend(product_keys)
                page_products[page_num] = set(product_keys)

                for p in result['rankingProducts']:
                    key = p['uniqueKey']
                    if key not in product_details:
                        product_details[key] = []

                    product_details[key].append({
                        'page': page_num,
                        'url': p['url'],
                        'name': p['name'],
                        'rank': p['rank']
                    })

        total_products = len(all_product_keys)
        unique_products = len(set(all_product_keys))
        duplicate_count = total_products - unique_products

        return {
            'total': total_products,
            'unique': unique_products,
            'duplicates': duplicate_count,
            'duplicate_rate': (duplicate_count / total_products * 100) if total_products > 0 else 0,
            'pageProducts': page_products,
            'productDetails': product_details
        }

    @staticmethod
    def print_duplicates(duplicate_stats):
        """Print duplicate statistics"""
        print("\n" + "="*80)
        print("[DUPLICATE CHECK - RANKING PRODUCTS (product_id + item_id + vendor_item_id)]")
        print("="*80)

        print(f"Total Ranking Products: {duplicate_stats['total']}")
        print(f"Unique Products: {duplicate_stats['unique']}")
        print(f"Duplicates: {duplicate_stats['duplicates']}")

        if duplicate_stats['duplicates'] > 0:
            print(f"\n[WARNING] {duplicate_stats['duplicates']}개 중복 상품 발견!")

            # Count occurrences
            product_counter = {}
            for page_products in duplicate_stats['pageProducts'].values():
                for key in page_products:
                    product_counter[key] = product_counter.get(key, 0) + 1

            # Find duplicates
            duplicates = [(key, count) for key, count in product_counter.items() if count > 1]

            print("\n[DUPLICATE PRODUCTS WITH FULL URLS]:")
            for key, count in duplicates:
                print(f"\n  Key {key}: {count}번 등장")

                # Find pages with this duplicate
                pages_with_duplicate = []
                for page_num, product_set in duplicate_stats['pageProducts'].items():
                    if key in product_set:
                        pages_with_duplicate.append(page_num)

                print(f"    Pages: {', '.join(map(str, pages_with_duplicate))}")

                # Print details
                details = duplicate_stats['productDetails'].get(key, [])
                for detail in details:
                    print(f"    Page {detail['page']}:")
                    print(f"      URL: {detail['url']}")
                    print(f"      Name: {detail['name'][:50]}...")
                    print(f"      Rank: {detail['rank']}")
        else:
            print("\n[OK] 중복 없음! 모든 상품이 고유합니다.")

        return duplicate_stats
